src/main/java/vg/civcraft/mc/civmodcore/dao/ManagedDatasource.java

```diff
@@ -0,0 +1,490 @@
+ *  3. Make sure you release Connections using {@link Connection#close()} as soon as you can (when done with them). 
+ *     a. Don't hold on to Connections. 
+ *     b. Close them. 
+ *     c. Use "try-with-resources" where-ever possible so that they are auto-closed.
+ *  4. If you have loops to insert a bunch of similar records, convert it to a batch. Find instructions in {@link #ManagedDatasource}.
+ *  5. If you have special needs like atomic multi-statement, do all your work on a single Connection and return it to 
+ *     a clean state when you are done. (turn auto-commit back on, ensure all transactions are committed, etc.)
+ *
+ * That should cover most cases. Note that points 2 & 3 are critical. Point 1 is required. Point 4 and 5 are highly recommended.
+ * 
+ * TODO: Make this instantiable via ConfigurationSerializable interface.
+ *
+ * @author ProgrammerDan
+ */
+public class ManagedDatasource {
+
+	private static final long MAX_WAIT_FOR_LOCK = 600000l;
+	private static final long WAIT_PERIOD = 500l;
+	
+	private ConnectionPool connections;
+	private Logger logger;
+	private ACivMod plugin;
+	
+	private static final String CHECK_CREATE_MIGRATIONS_TABLE = 
+			"CREATE TABLE IF NOT EXISTS managed_plugin_data ("
+					+ "managed_id BIGINT NOT NULL AUTO_INCREMENT, "
+					+ "plugin_name VARCHAR(120) NOT NULL, "
+					+ "management_began TIMESTAMP NOT NULL DEFAULT NOW(), "
+					+ "current_migration_number INT NOT NULL, "
+					+ "last_migration TIMESTAMP, "
+					+ "CONSTRAINT pk_managed_plugin_data PRIMARY KEY (managed_id), "
+					+ "CONSTRAINT uniq_managed_plugin UNIQUE (plugin_name), "
+					+ "INDEX idx_managed_plugin USING BTREE (plugin_name)" +
+					");";
+	
+	private static final String CHECK_LAST_MIGRATION = 
+			"SELECT current_migration_number FROM managed_plugin_data WHERE plugin_name = ?;";
+	
+	private static final String RECORD_MIGRATION =
+			"INSERT INTO managed_plugin_data (plugin_name, current_migration_number, last_migration) " 
+					+ "VALUES (?, ?, NOW()) ON DUPLICATE KEY UPDATE plugin_name = VALUES(plugin_name), "
+					+ "current_migration_number = VALUES(current_migration_number), "
+					+ "last_migration = VALUES(last_migration);";
+	
+	private static final String CHECK_CREATE_LOCK_TABLE = 
+			"CREATE TABLE IF NOT EXISTS managed_plugin_locks ("
+					+ "plugin_name VARCHAR(120) NOT NULL, "
+					+ "lock_time TIMESTAMP NOT NULL DEFAULT NOW(), "
+					+ "CONSTRAINT pk_managed_plugin_locks PRIMARY KEY (plugin_name)" +
+					");";
+	
+	private static final String CLEANUP_LOCK_TABLE = 
+			"DELETE FROM managed_plugin_locks WHERE lock_time <= TIMESTAMPADD(HOUR, -8, NOW());";
+	
+	private static final String ACQUIRE_LOCK = 
+			"INSERT IGNORE INTO managed_plugin_locks (plugin_name) VALUES (?);";
+	
+	private static final String RELEASE_LOCK =
+			"DELETE FROM managed_plugin_locks WHERE plugin_name = ?";
+
+	private int first_migration;
+	private TreeMap<Integer, LinkedList<String>> migrations;
```

Comment:
Instead of having 3 maps of the same type using the same key to keep synchronized why not just create an inner data class to hold the migration commands, ignore flag, and post-migration callables? It's more flexible in the future if additional state needs to be tracked.